File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.py
# take the dimension of the grid
# if the cell at the top-left or bottom right is not equal to a 0, return -1
# declare a variable queue that will hold the cordinates and the path of the elements in the traversal
# while the queue is not empty, take the length of the queue
# in the range length
# pop out the current cordinates and path length from the queue
# if the current cordinates point to the bottom right cell return the path
# add the cordinates and path length of all the valid adjacent cells to the queue
# increment the path length outside of the for loop
# return -1 outside the while loop

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        if grid[0][0] != 0 or grid[-1][-1] != 0:
            return -1
        queue = [(0, 0, 1)]
        visited = set()

        def isValid(i, j):
            if i < 0 or j < 0 or i == n or j == n or grid[i][j] != 0 or (i, j) in visited:
                return False
            return True

        directions = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1] ]
        while queue:
            size = len(queue)

            for _ in range(size):
                r, c, path_length = queue.pop(0)
                if r == len(grid) -1 and c == len(grid[0]) -1:
                    return path_length
                for dr, dc in directions:
                    row = dr + r
                    column = dc + c

                    if (isValid(row, column)):
                        queue.append([row, column, path_length + 1])
                        visited.add((row, column))
                    
        return -1
                

                

        

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.js
var shortestPathBinaryMatrix = function(grid) {
    let n = grid.length;
    let visited = new Set();
    if (grid[0][0] !== 0 || grid[n - 1][n - 1] !== 0) return -1;
    let queue = [[0, 0, 1]]; // Initialize with starting point and path length

    function isValid(i, j) {
        // Correct the bounds checking and the visited check
        return i >= 0 && j >= 0 && i < n && j < n && !visited.has(i + ',' + j);
    }

    let directions = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]];

    while (queue.length) {
        let size = queue.length;

        for (let i = 0; i < size; i++) {
            let [x, y, path] = queue.shift();
            if (x === n - 1 && y === n - 1) return path; // Found the end point
            for (let [a, b] of directions) {
                let new_x = a + x;
                let new_y = b + y;

                if (isValid(new_x, new_y) && grid[new_x][new_y] !== 1) {
                    visited.add(new_x + ',' + new_y); // Use new_x and new_y
                    queue.push([new_x, new_y, path + 1]); // Push the new coordinates and updated path length
                }
            }
        }
    }
    return -1; // No path found
};